h2quic/response_writer.go

```diff
@@ -86,12 +98,129 @@ func (w *responseWriter) Flush() {}
+	if w.headerStream.StreamID()%2 == 0 {
+		return http2.ErrRecursivePush
+	}
+	// Default options. Copied from net/http2/server.go
+	if opts.Method == "" {
+		opts.Method = "GET"
+	}
+	if opts.Header == nil {
+		opts.Header = http.Header{}
+	}
+	if opts.Method != "GET" && opts.Method != methodHEAD {
+		return fmt.Errorf("method %q must be GET or HEAD", opts.Method)
+	}
+	// Validate the target. Copied from net/http2/server.go
+	u, err := url.Parse(target)
+	if err != nil {
+		return err
+	}
+	wantScheme := "https"
+	if u.Scheme == "" {
+		if !strings.HasPrefix(target, "/") {
+			return fmt.Errorf("target must be an absolute URL or an absolute path: %q", target)
+		}
+		u.Scheme = wantScheme
+		u.Host = w.requestHost
+	} else {
+		if u.Scheme != wantScheme {
+			return fmt.Errorf("cannot push URL with scheme %q from request with scheme %q", u.Scheme, wantScheme)
+		}
+		if u.Host == "" {
+			return errors.New("URL must have a host")
+		}
+	}
+	for k := range opts.Header {
+		if strings.HasPrefix(k, ":") {
+			return fmt.Errorf("promised request headers cannot include pseudo header %q", k)
+		}
+		switch strings.ToLower(k) {
+		case "content-length", "content-encoding", "trailer", "te", "expect", "host":
+			return fmt.Errorf("promised request headers cannot include %q", k)
+		}
+	}
+	if err = checkValidHTTP2RequestHeaders(opts.Header); err != nil {
+		return err
+	}
+	authority := u.Host
+	path := u.RequestURI()
+	contentLengthStr := "0" // a request does not have content
+	// Construct HTTP headers for request in push promise
+	var headers bytes.Buffer
+	enc := hpack.NewEncoder(&headers)
+	enc.WriteField(hpack.HeaderField{Name: ":method", Value: opts.Method})
+	enc.WriteField(hpack.HeaderField{Name: ":path", Value: path})
+	enc.WriteField(hpack.HeaderField{Name: ":authority", Value: authority})
+	enc.WriteField(hpack.HeaderField{Name: ":scheme", Value: u.Scheme})
+	for k, v := range opts.Header {
+		for index := range v {
+			enc.WriteField(hpack.HeaderField{Name: strings.ToLower(k), Value: v[index]})
+		}
+	}
+	// Get new data stream
+	newDataStream, err := w.session.OpenStream()
+	if err != nil {
+		if err == qerr.TooManyOpenStreams {
+			err = http2.ErrPushLimitReached
+		}
+		return err
+	}
+	if newDataStream == nil {
+		return fmt.Errorf("Got invalid data stream")
+	}
+	newDataStreamID := newDataStream.StreamID()
+	if newDataStreamID%2 != 0 {
+		return fmt.Errorf("Cannot push on client side stream")
+	}
+	// Write push promise header
+	p := http2.PushPromiseParam{
+		StreamID:      uint32(w.dataStreamID),
+		PromiseID:     uint32(newDataStreamID),
+		BlockFragment: headers.Bytes(),
+		EndHeaders:    true,
+	}
+	headerFramer := http2.NewFramer(w.headerStream, nil)
+
+	utils.Debugf("Sending PUSH_PROMISE for target '%s' on stream %d, promised on stream %d", target, w.headerStream.StreamID(), newDataStreamID)
+
+	w.headerStreamMutex.Lock()
+	err = headerFramer.WritePushPromise(p)
```

Comment:
Nit: Don't save the `PushPromiseParam` to `p`, write it directly (like we do it with the `HeadersFrameParam`).